code/modules/wiremod/components/action/pathfind.dm

```diff
@@ -0,0 +1,124 @@
+/**
+ * # Pathfinding component
+ *
+ * Calcualtes a path, returns a list of entities. Each entity is the next step in the path. Can be used with the direction component to move.
+ */
+/obj/item/circuit_component/pathfind
+	display_name = "Pathfinder"
+	display_desc = "When triggered, the next step to the target's location as an entity. This can be used with the direction component and the drone shell to make it move on its own. The Id Card input port is for considering ID access when pathing, it does not give the shell actual access."
+	circuit_flags = CIRCUIT_FLAG_INPUT_SIGNAL|CIRCUIT_FLAG_OUTPUT_SIGNAL
+
+	var/datum/port/input/input_X
+	var/datum/port/input/input_Y
+	var/datum/port/input/id_card
+
+	var/datum/port/output/output
+	var/datum/port/output/finished
+	var/datum/port/output/failed
+	var/datum/port/output/reason_failed
+
+	var/list/path
+	var/turf/old_dest
+	var/turf/next_turf
+
+	// Cooldown to limit how frequently we can path to the same location.
+	var/same_path_cooldown = 5 SECONDS
+	var/different_path_cooldown = 30 SECONDS
+
+/obj/item/circuit_component/pathfind/get_ui_notices()
+	. = ..()
+	// Not necessary to show the same path cooldown, since it doesn't change much for the player
+	. += create_ui_notice("Pathfinding Cooldown: [DisplayTimeText(different_path_cooldown)]", "orange", "stopwatch")
+
+/obj/item/circuit_component/pathfind/Initialize()
+	. = ..()
+	input_X = add_input_port("Target X", PORT_TYPE_NUMBER, FALSE)
+	input_Y = add_input_port("Target Y", PORT_TYPE_NUMBER, FALSE)
+	id_card = add_input_port("ID Card", PORT_TYPE_ATOM, FALSE)
+
+	output = add_output_port("Next step", PORT_TYPE_ATOM)
+	finished = add_output_port("Arrived to destination", PORT_TYPE_SIGNAL)
+	failed = add_output_port("Failed", PORT_TYPE_SIGNAL)
+	reason_failed = add_output_port("Fail reason", PORT_TYPE_STRING)
+
+/obj/item/circuit_component/pathfind/Destroy()
+	input_X = null
+	input_Y = null
+	id_card = null
+
+	output = null
+	finished = null
+	failed = null
+	reason_failed = null
+
+	path = null
+	old_dest = null
+	next_turf = null
+	return ..()
+
+/obj/item/circuit_component/pathfind/input_received(datum/port/input/port)
+	. = ..()
+	if(.)
+		return
+
+	var/target_X = input_X.input_value
+	if(isnull(target_X))
+		return
+
+	var/target_Y = input_Y.input_value
+	if(isnull(target_Y))
+		return
+
+	var/atom/path_id = id_card.input_value
+	if(path_id && !istype(path_id, /obj/item/card/id))
+		path_id = null
+		failed.set_output(COMPONENT_SIGNAL)
+		reason_failed.set_output("Object marked is not an ID! Using no ID instead.")
```

Comment:
This shouldn't really flag up a fail if it doesn't stop execution.

Reply:
Yeah I'm conflicted on this. If the player accidentally selects an atom that's not an ID card I want them to know so, but if I straight up stop execution then I basically force the player to scan any ID card for the component to work, since you can't clear the entity input. This would suck for people who change their minds and decide that the pathfinding shouldn't take access into consideration, so it's either warn them with a fail and don't stop, or don't warn them at all, which would make em confused.